result/result.py

```diff
@@ -1,75 +1,137 @@
 # -*- coding: utf-8 -*-
 from __future__ import print_function, division, absolute_import, unicode_literals
+from typing import Generic, TypeVar, Union, Any, Optional, cast, overload
 
 
-class Result(object):
+E = TypeVar("E")
+T = TypeVar("T")
+A = TypeVar("A")
+
+
+class Result(Generic[E, T]):
     """
     A simple `Result` type inspired by Rust.
 
     Not all methods (https://doc.rust-lang.org/std/result/enum.Result.html)
     have been implemented, only the ones that make sense in the Python context.
-    You still don't get any type checking done.
-
     """
-    def __init__(self, force=False):
+    def __init__(self, is_ok, value, force=False):
+        # type: (bool, Union[E, T], bool) -> None
+        """Do not call this constructor, use the Ok or Err class methods instead.
+
+        There are no type guarantees on the value if this is called directly.
+
+        Args:
+            is_ok: If this represents an ok result
+            value: The value inside the result
+            force: Force creation of the object. This is false by default to prevent accidentally
+                creating instance of a Result in an unsafe way.
+        """
         if force is not True:
             raise RuntimeError("Don't instantiate a Result directly. "
                                "Use the Ok(value) and Err(error) class methods instead.")
+        else:
+            self._is_ok = is_ok
+            self._value = value
 
     def __eq__(self, other):
-        return self._type == other._type and self._val == other._val
+        # type: (Any) -> bool
+        return cast(bool, (self.__class__ == other.__class__ and
+                           self.is_ok() == other.is_ok() and
+                           self._value == other._value))
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+        return not (self == other)
```

Comment:
I believe so. I added a test that fails if I comment this function out, and the tests currently fail if I comment the hash function. It looks like this may have changed in python 3 though. [This page](https://docs.python.org/2/reference/datamodel.html#object.__ne__) says to implement both, but (the python 3 docs)[https://docs.python.org/3/reference/datamodel.html#object.__ne__] say that `__ne__` delegates to `__eq__` by default.